File: iosApp/iosApp/Features/Feed/FeedViewModel.swift
import Foundation
import Shared

@MainActor
final class FeedViewModelOwner: ObservableObject {
    @Published var feedItems: [FeedItem] = []
    @Published var isLoading = true
    @Published var isRefreshing = false
    @Published var errorMessage: String?
    @Published var unreadCount: Int32 = 0

    private let vm: FeedViewModel
    private let store: ViewModelStore

    init() {
        store = ViewModelStore()
        vm = KoinHelper.shared.createFeedViewModel()
        store.put(key: "FeedViewModel", viewModel: vm)
    }

    deinit { store.clear() }

    func observeUiState() async {
        for await state in vm.uiState {
            feedItems = state.feedItems as? [FeedItem] ?? []
            isLoading = state.isLoading
            isRefreshing = state.isRefreshing
            errorMessage = state.error
            unreadCount = state.unreadCount
        }
    }

    func refresh() { vm.refresh() }
    func markAsRead() { vm.markAsRead() }
}

File: shared/src/commonMain/kotlin/com/riox432/civitdeck/presentation/feed/FeedViewModel.kt
package com.riox432.civitdeck.presentation.feed

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.riox432.civitdeck.domain.model.FeedItem
import com.riox432.civitdeck.domain.usecase.GetCreatorFeedUseCase
import com.riox432.civitdeck.domain.usecase.GetUnreadFeedCountUseCase
import com.riox432.civitdeck.domain.usecase.MarkFeedReadUseCase
import com.riox432.civitdeck.domain.util.suspendRunCatching
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

data class FeedUiState(
    val feedItems: List<FeedItem> = emptyList(),
    val isLoading: Boolean = true,
    val isRefreshing: Boolean = false,
    val error: String? = null,
    val unreadCount: Int = 0,
)

class FeedViewModel(
    private val getCreatorFeedUseCase: GetCreatorFeedUseCase,
    private val getUnreadFeedCountUseCase: GetUnreadFeedCountUseCase,
    private val markFeedReadUseCase: MarkFeedReadUseCase,
) : ViewModel() {

    private val _uiState = MutableStateFlow(FeedUiState())
    val uiState: StateFlow<FeedUiState> = _uiState

    init {
        loadFeed()
        observeUnreadCount()
    }

    fun refresh() {
        loadFeed(forceRefresh = true)
    }

    fun markAsRead() {
        viewModelScope.launch {
            markFeedReadUseCase()
        }
    }

    private fun loadFeed(forceRefresh: Boolean = false) {
        viewModelScope.launch {
            _uiState.update {
                it.copy(
                    isLoading = !forceRefresh && it.feedItems.isEmpty(),
                    isRefreshing = forceRefresh,
                    error = null,
                )
            }
            suspendRunCatching { getCreatorFeedUseCase(forceRefresh) }
                .onSuccess { items ->
                    _uiState.update {
                        it.copy(
                            feedItems = items,
                            isLoading = false,
                            isRefreshing = false,
                        )
                    }
                    if (items.isNotEmpty()) markFeedReadUseCase()
                }
                .onFailure { e ->
                    _uiState.update {
                        it.copy(
                            isLoading = false,
                            isRefreshing = false,
                            error = e.message ?: "Failed to load feed",
                        )
                    }
                }
        }
    }

    private fun observeUnreadCount() {
        viewModelScope.launch {
            getUnreadFeedCountUseCase().collect { count ->
                _uiState.update { it.copy(unreadCount = count) }
            }
        }
    }
}
